File: test_programs/python/freegroups/objects/Morphism.py
from .FreeGroup import *

class Morphism:
    morphism_map : dict[Symbol, Word]
    free_group : FreeGroup

    def __init__(self, group : FreeGroup, map : dict[Symbol, Word]) -> None:
        self.free_group = group
        for sym in group.basis:
            if sym not in map:
                map[sym] = Word((Elem(sym, 1),))
        self.morphism_map = map

    def map(self, word : Word) -> Word:
        newWord = MutableWord([])
        for elem in word.word:
            newWord.word.extend((self.morphism_map[elem.sym] ** elem.exp).word)
        return newWord.immutable().reduced()

File: test_programs/haskell/Objects/Morphism.hs
module Morphism (
    Morphism,
    mkMorphism,
    amap,
)
where

import Data.Group
import Data.Map qualified as M
import Element
import FreeGroup
import GroupWord
import ReducedWord

data Morphism = Morphism
    { group :: FreeGroup
    , mmap :: M.Map Symbol RWord
    }
    deriving (Show)

mkMorphism :: FreeGroup -> [(Symbol, RWord)] -> Morphism
mkMorphism group mapList = Morphism group (makeMap group mapList)

makeMap :: FreeGroup -> [(Symbol, RWord)] -> M.Map Symbol RWord
makeMap group mapList = foldl mapHelper M.empty (symbols group)
  where
    mapHelper currMap sym = M.insert sym (get sym) currMap
    get sym = case lookup sym mapList of
        Just word -> word
        Nothing -> elemRWord (Elem sym 1)

amap :: Morphism -> GroupWord -> RWord
amap morph [] = RWord []
amap morph ((Elem sym expo) : rest) = pow (getWord sym) expo <> amap morph rest
  where
    getWord sym = mmap morph M.! sym
